### ISSUE TITLE
[Feature Request] Add user defined callbacks.

### ISSUE THREAD

**Issue by jakubczakon** [pfnet/optuna#389]
First of all amazing job!
Thank you so much for this library.

Secondly, I think it would be a good idea to have an option to define callbacks.
For example, I would like to log the progress live to [Neptune](neptune.ml) tracking tool.

I made this [example experiment](https://ui.neptune.ml/jakub-czakon/blog-hpo/e/BLOG-215/charts) public so you can go there if you want. Full script is [available here](https://ui.neptune.ml/jakub-czakon/blog-hpo/e/BLOG-215/source-code?path=.&file=search_optuna.py).
 
![image](https://gist.githubusercontent.com/jakubczakon/f754769a39ea6b8fa9728ede49b9165c/raw/5c0fee0a92d3f32157c1d01e867e9a0ad53b0573/optuna_search.png)

Anyhow, what I ended up doing was:

```python
def neptune_monitor(params, score):
    neptune.send_metric('run_score', score)
    neptune.send_text('run_parameters', str(params))

def objective(trial):
    params = OrderedDict([('learning_rate',trial.suggest_loguniform('learning_rate', 0.01, 0.5)),
                        ('max_depth',trial.suggest_int('max_depth', 1, 30)),
                        ('num_leaves',trial.suggest_int('num_leaves', 2, 100)),
                        ('min_data_in_leaf',trial.suggest_int('min_data_in_leaf', 10, 1000)),
                        ('feature_fraction',trial.suggest_uniform('feature_fraction', 0.1, 1.0)),
                        ('subsample',trial.suggest_uniform('subsample', 0.1, 1.0))
                        ])
    score = -1.0 * train_evaluate(X, y, params)
    neptune_monitor(params, score)
    return score
```

It gets the job done but definitely is not ideal.

I think a good place to register callbacks would be right after the study is created. 
I would just have access to trial object post iteration and I could get everything I need from there.
For example,

```python

def neptune_monitor(trial):
    trial_df = trial.study.trials_dataframe().iloc[-1]
    neptune.send_metric('run_score', trial_df['value']   )
    neptune.send_text('run_parameters', str(trial_df['params']))
...
study = optuna.create_study()  
study.register_callbacks([neptune_monitor])
study.optimize(objective, n_trials=100)

```

What do you think?

**Comment by sile** [comment]
Thank you for your request and detailed explanation!

We have received similar feature requests from other people, and are planning to implement it. The design and implementation details have not been decided yet, but we will create a draft PR in the near future.
We would very appreciate it if you could give us some feedback when the PR is created.

**Comment by jakubczakon** [comment]
Thank you @sile, I would love to help!

On a separate note, I am actually writing a blog series about HPO in python and Optuna is going to be part 5 of that.
[Part 0](https://towardsdatascience.com/hyperparameter-optimization-in-python-part-0-introduction-c4b66791614b) and [Part 1](https://towardsdatascience.com/hyperparameter-optimization-in-python-part-1-scikit-optimize-754e485d24fe) are already published on towards data science.

**Comment by sile** [comment]
>  I am actually writing a blog series about HPO in python and Optuna is going to be part 4 of that.

Great! We are looking forward to reading it!

**Comment by yonatanMedan** [comment]
i wrote a simple callback implementation for my personal project using optuna:

```python
class Experiment:
    def __init__(self,study_name,storage,objective,callbacks =[]):
        self.study = optuna.create_study(study_name=study_name,storage=storage,load_if_exists=True) 
        self.callbacks = callbacks
        
    def step(self,trial):
        error = objective(trial)
        self.after_error(error)
        return error
    
    def after_error(self,error):
        for callback in self.callbacks:
            if hasattr(callback,"after_error"):
                callback.after_error(self,error)
                
    def optimize(self,n_trials = 10):
        self.study.optimize(self.step,n_trials)
```
example callback:
```python
class VisualizeTrialsCallback:
    def after_error(self,experiment,loss,**kwargs):
        try:
            plot_triales(experiment.study)
        except Exception:
            pass

```
and then you can do:
```python
experiment = Experiment(study_name=name,storage=stotage_name,objective=objective,callbacks=[VisualizeTrialsCallback()])
```
what do you think?

**Comment by sile** [comment]
@yonatanMedan 

Thank you for sharing your idea.
I think that it's a nice way to enable callbacks without modifying Optuna's code.

FYI, we're going to implement the feature to Optuna at #480, because it's a frequently requested feature by users.
Note that the PR is a still draft and the API design may change before merging.

**Comment by sile** [comment]
I close this issue because this feature was implemented in #480.
Feel free to re-open this issue if you have further concerns.
